tcptracer-bpf.c

```diff
@@ -0,0 +1,961 @@
+	updated_status.offset_ino = status->offset_ino;
+	updated_status.offset_family = status->offset_family;
+	updated_status.offset_daddr_ipv6 = status->offset_daddr_ipv6;
+	updated_status.err = 0;
+	updated_status.saddr = status->saddr;
+	updated_status.daddr = status->daddr;
+	updated_status.sport = status->sport;
+	updated_status.dport = status->dport;
+	updated_status.netns = status->netns;
+	updated_status.family = status->family;
+
+	int i;
+	for (i = 0; i < 4; i++) {
+		updated_status.daddr_ipv6[i] = status->daddr_ipv6[i];
+	}
+
+	u32 possible_saddr;
+	u32 possible_daddr;
+	u16 possible_sport;
+	u16 possible_dport;
+	possible_net_t *possible_skc_net;
+	u32 possible_netns;
+	u16 possible_family;
+	long ret = 0;
+
+	switch (status->what) {
+		case GUESS_SADDR:
+			possible_saddr = 0;
+			bpf_probe_read(&possible_saddr, sizeof(possible_saddr), ((char *)skp) + status->offset_saddr);
+			updated_status.saddr = possible_saddr;
+			break;
+		case GUESS_DADDR:
+			possible_daddr = 0;
+			bpf_probe_read(&possible_daddr, sizeof(possible_daddr), ((char *)skp) + status->offset_daddr);
+			updated_status.daddr = possible_daddr;
+			break;
+		case GUESS_FAMILY:
+			possible_family = 0;
+			bpf_probe_read(&possible_family, sizeof(possible_family), ((char *)skp) + status->offset_family);
+			updated_status.family = possible_family;
+			break;
+		case GUESS_SPORT:
+			possible_sport = 0;
+			bpf_probe_read(&possible_sport, sizeof(possible_sport), ((char *)skp) + status->offset_sport);
+			updated_status.sport = possible_sport;
+			break;
+		case GUESS_DPORT:
+			possible_dport = 0;
+			bpf_probe_read(&possible_dport, sizeof(possible_dport), ((char *)skp) + status->offset_dport);
+			updated_status.dport = possible_dport;
+			break;
+		case GUESS_NETNS:
+			possible_netns = 0;
+			possible_skc_net = NULL;
+			bpf_probe_read(&possible_skc_net, sizeof(possible_net_t *), ((char *)skp) + status->offset_netns);
+			// if we get a kernel fault, it means possible_skc_net
+			// is an invalid pointer, signal an error so we can go
+			// to the next offset_netns
+			ret = bpf_probe_read(&possible_netns, sizeof(possible_netns), ((char *)possible_skc_net) + status->offset_ino);
+			if (ret == -EFAULT) {
+				updated_status.err = 1;
+				break;
+			}
+			updated_status.netns = possible_netns;
+			break;
+		default:
+			// not for us
+			return 0;
+	}
+
+	bpf_map_update_elem(&tcptracer_status, &zero, &updated_status, BPF_ANY);
+
+	return 0;
+}
+
+static int are_offsets_ready_v6(struct tcptracer_status_t *status, struct sock *skp, u64 pid) {
+	u64 zero = 0;
+
+	switch (status->status) {
+		case TCPTRACER_STATUS_UNINITIALIZED:
+			return 0;
+		case TCPTRACER_STATUS_CHECKING:
+			break;
+		case TCPTRACER_STATUS_CHECKED:
+			return 0;
+		case TCPTRACER_STATUS_READY:
+			return 1;
+		default:
+			return 0;
+	}
+
+	if (status->pid_tgid >> 32 != pid >> 32)
+		return 0;
+
+	struct tcptracer_status_t updated_status = { };
+	updated_status.status = TCPTRACER_STATUS_CHECKED;
```

Comment:
Unfortunately in some old kernels (like 4.4), declaring and initializing variables in the same "line" makes the kernel verifier reject the program.